src/messages.cpp

```diff
@@ -265,170 +265,586 @@ Sender::sender_type() const
+    case SenderType::member: {
+      auto candidate = membership_mac(suite, opt::get(membership_key), context);
+      if (candidate != opt::get(membership_tag)) {
+        return {};
+      }
+      break;
+    }
+
+    default:
+      break;
   }
 
-  return tag == opt::get(membership_tag);
+  return MLSMessageContentAuth{
+    WireFormat::mls_plaintext,
+    content,
+    auth,
+  };
+}
+
+MLSPlaintext::MLSPlaintext(MLSMessageContentAuth content_auth)
+  : content(std::move(content_auth.content))
+  , auth(std::move(content_auth.auth))
+{
+  if (content_auth.wire_format != WireFormat::mls_plaintext) {
+    throw InvalidParameterError("Wire format mismatch (not mls_plaintext)");
+  }
+}
+
+bytes
+MLSPlaintext::membership_mac(CipherSuite suite,
+                             const bytes& membership_key,
+                             const std::optional<GroupContext>& context) const
+{
+  auto tbm = tls::marshal(MLSMessageContentTBM{
+    { WireFormat::mls_plaintext, content, context },
+    auth,
+  });
+
+  return suite.digest().hmac(membership_key, tbm);
+}
+
+tls::ostream&
+operator<<(tls::ostream& str, const MLSPlaintext& obj)
+{
+  str << obj.content << obj.auth;
+
+  if (obj.content.sender.sender_type() == SenderType::member) {
+    str << opt::get(obj.membership_tag);
+  }
+
+  return str;
+}
+
+tls::istream&
+operator>>(tls::istream& str, MLSPlaintext& obj)
+{
+  str >> obj.content;
+
+  obj.auth.content_type = obj.content.content_type();
+  str >> obj.auth;
+
+  if (obj.content.sender.sender_type() == SenderType::member) {
+    obj.membership_tag.emplace();
+    str >> opt::get(obj.membership_tag);
+  }
+
+  return str;
+}
+
+// struct {
+//     select (MLSCiphertext.content_type) {
+//         case application:
+//           opaque application_data<V>;
+//
+//         case proposal:
+//           Proposal proposal;
+//
+//         case commit:
+//           Commit commit;
+//     }
+//
+//     MLSMessageAuth auth;
+//     opaque padding<V>;
+// } MLSCiphertextContent;
+static bytes
+marshal_ciphertext_content(const MLSMessageContent& content,
+                           const MLSMessageAuth& auth,
+                           size_t padding_size)
+{
+  auto w = tls::ostream{};
+  var::visit([&w](const auto& val) { w << val; }, content.content);
+  w << auth << bytes(padding_size, 0);
+  return w.bytes();
+}
+
+static void
+unmarshal_ciphertext_content(const bytes& content_pt,
+                             MLSMessageContent& content,
+                             MLSMessageAuth& auth)
+{
+  auto r = tls::istream(content_pt);
+
+  switch (auth.content_type) {
+    case ContentType::application:
+      content.content.emplace<ApplicationData>();
+      break;
+
+    case ContentType::proposal:
+      content.content.emplace<Proposal>();
+      break;
+
+    case ContentType::commit:
+      content.content.emplace<Commit>();
+      break;
+
+    default:
+      throw InvalidParameterError("Unknown content type");
+  }
```

Comment:
Do this switch in `unprotect`.  Then the `visit()` here will have the intended effect.